File: mobile-engage/src/main/java/com/emarsys/mobileengage/iam/InAppStartAction.java
package com.emarsys.mobileengage.iam;

import android.app.Activity;

import com.emarsys.core.activity.ActivityLifecycleAction;
import com.emarsys.core.storage.Storage;
import com.emarsys.core.util.Assert;
import com.emarsys.mobileengage.event.EventServiceInternal;

public class InAppStartAction implements ActivityLifecycleAction {

    private EventServiceInternal eventServiceInternal;
    private Storage<String> contactTokenStorage;

    public InAppStartAction(EventServiceInternal eventServiceInternal, Storage<String> contactTokenStorage) {
        Assert.notNull(eventServiceInternal, "EventServiceInternal must not be null!");
        Assert.notNull(contactTokenStorage, "ContactTokenStorage must not be null!");

        this.eventServiceInternal = eventServiceInternal;
        this.contactTokenStorage = contactTokenStorage;
    }

    @Override
    public void execute(Activity activity) {
        if (contactTokenStorage.get() != null) {
            eventServiceInternal.trackInternalCustomEvent("app:start", null, null);
        }
    }
}

File: mobile-engage/src/main/java/com/emarsys/mobileengage/iam/InAppStartAction.kt
package com.emarsys.mobileengage.iam

import android.app.Activity
import com.emarsys.core.activity.ActivityLifecycleAction
import com.emarsys.core.api.proxyApi
import com.emarsys.core.storage.Storage
import com.emarsys.core.util.log.Logger.Companion.info
import com.emarsys.core.util.log.entry.AppEventLog
import com.emarsys.mobileengage.di.mobileEngage
import com.emarsys.mobileengage.event.EventServiceInternal

class InAppStartAction(private val eventServiceInternal: EventServiceInternal, private val contactTokenStorage: Storage<String?>) : ActivityLifecycleAction {

    override fun execute(activity: Activity?) {
        val coreSdkHandler = mobileEngage().coreSdkHandler
        coreSdkHandler.post {
            if (contactTokenStorage.get() != null) {
                eventServiceInternal.proxyApi(coreSdkHandler).trackInternalCustomEventAsync("app:start", null, null)
            }
            info(AppEventLog("app:start", null))
        }

    }
}